File: internal/panel/mockdata.go
package panel

import (
	"fmt"

	"github.com/wrfly/deepseek-panel/internal/deepseek"
)

// MockData 本地模拟数据：结构与真实接口一致，按小时生成、按天变化，
// 用于离线测试，避免被平台 WAF 限流干扰。
// 与 Swift MockData 对应。
type MockData struct{}

// MockFetch 返回模拟的 summary / keys / amount / cost。
func (MockData) Fetch(window StatsWindow, tz int) (*deepseek.UserSummary, []deepseek.APIKeyInfo, *deepseek.UsageAmountData, *deepseek.CostData) {
	keys := []deepseek.APIKeyInfo{
		{CreatedAt: intPtr(1_700_000_000), TrackingID: "mock-claude", SensitiveID: "sk-mock-1***", Name: "claude"},
		{CreatedAt: intPtr(1_700_000_000), TrackingID: "mock-codex", SensitiveID: "sk-mock-2***", Name: "codex"},
		{CreatedAt: intPtr(1_700_000_000), TrackingID: "mock-cursor", SensitiveID: "sk-mock-3***", Name: "cursor"},
		{CreatedAt: intPtr(1_700_000_000), TrackingID: "mock-wechat", SensitiveID: "sk-mock-4***", Name: "wechat filter"},
	}
	models := []string{
		"deepseek-chat & deepseek-reasoner",
		"deepseek-v4-flash",
		"deepseek-v4-pro",
	}

	var amountSeries []deepseek.AmountSeries
	var cnySeries, usdSeries []deepseek.CostSeries

	for _, key := range keys {
		var keyModels []string
		switch key.Name {
		case "claude":
			keyModels = []string{"deepseek-v4-flash", "deepseek-v4-pro"}
		case "codex":
			keyModels = []string{"deepseek-chat & deepseek-reasoner", "deepseek-v4-pro"}
		case "wechat filter":
			keyModels = []string{"deepseek-v4-flash"}
		default:
			continue // cursor 不活跃
		}

		for _, model := range keyModels {
			var amountBuckets []deepseek.AmountBucket
			var cnyBuckets, usdBuckets []deepseek.CostBucket
			for t := window.RequestStart; t < window.RequestEnd; t += 3600 {
				usage := mockUsage(t, key.Name, model)
				amountBuckets = append(amountBuckets, deepseek.AmountBucket{Time: int(t), Usage: usage})
				usd := mockCostUSD(usage, model)
				cnyBuckets = append(cnyBuckets, deepseek.CostBucket{Time: int(t), Cost: fmt.Sprintf("%.10f", usd*7.2)})
				usdBuckets = append(usdBuckets, deepseek.CostBucket{Time: int(t), Cost: fmt.Sprintf("%.10f", usd)})
			}

			seriesKey := deepseek.SeriesKey{
				TrackingID:  key.TrackingID,
				Name:        key.Name,
				SensitiveID: key.SensitiveID,
				Valid:       boolPtr(true),
			}
			amountSeries = append(amountSeries, deepseek.AmountSeries{APIKey: seriesKey, Model: model, Buckets: amountBuckets})
			cnySeries = append(cnySeries, deepseek.CostSeries{APIKey: seriesKey, Model: model, Buckets: cnyBuckets})
			usdSeries = append(usdSeries, deepseek.CostSeries{APIKey: seriesKey, Model: model, Buckets: usdBuckets})
		}
	}

	amount := &deepseek.UsageAmountData{
		Start:  int(window.RequestStart),
		End:    int(window.RequestEnd),
		Bucket: 3600,
		Models: &models,
		Series: amountSeries,
	}
	cost := &deepseek.CostData{
		Start:  int(window.RequestStart),
		End:    int(window.RequestEnd),
		Bucket: 3600,
		Models: &models,
		Data: &[]deepseek.CurrencyCostSeries{
			{Currency: "CNY", Series: cnySeries},
			{Currency: "USD", Series: usdSeries},
		},
	}
	summary := &deepseek.UserSummary{
		NormalWallets: []deepseek.Wallet{
			{Currency: "USD", Balance: "8.4200000000", TokenEstimation: strPtr("0")},
			{Currency: "CNY", Balance: "3.1400000000", TokenEstimation: strPtr("0")},
		},
		BonusWallets: []deepseek.Wallet{
			{Currency: "USD", Balance: "0", TokenEstimation: strPtr("0")},
		},
		TotalCosts: []deepseek.CostTotal{
			{Currency: "USD", Amount: "0.8800000000"},
			{Currency: "CNY", Amount: "1.2300000000"},
		},
	}
	return summary, keys, amount, cost
}

func mockUsage(time int64, keyName, model string) deepseek.TokenUsage {
	hour := (time % 86400) / 3600
	busy := 0.2
	if hour >= 9 && hour <= 22 {
		busy = 1.0
	}
	keyFactor := 1.0
	if keyName == "wechat filter" {
		keyFactor = 0.04
	}
	modelFactor := 1.0
	if model == "deepseek-v4-pro" {
		modelFactor = 2.0
	}
	base := int(2_000_000 * busy * keyFactor * modelFactor)
	salt := fnv1a(keyName)

	hit := mockRand(time, int64(salt+1), max(base, 1))
	miss := mockRand(time, int64(salt+2), max(base/8, 500))
	response := mockRand(time, int64(salt+3), max(base/20, 100))
	requests := mockRand(time, int64(salt+4), max(base/6000, 1))

	return deepseek.TokenUsage{
		Request:              &requests,
		ResponseToken:        &response,
		PromptCacheHitToken:  &hit,
		PromptCacheMissToken: &miss,
	}
}

func mockCostUSD(usage deepseek.TokenUsage, model string) float64 {
	hit := float64(intOrZero(usage.PromptCacheHitToken))
	miss := float64(intOrZero(usage.PromptCacheMissToken))
	output := float64(intOrZero(usage.ResponseToken))
	prices := struct{ hit, miss, output float64 }{0.02, 1.0, 2.0}
	if model == "deepseek-v4-pro" {
		prices = struct{ hit, miss, output float64 }{0.025, 3.0, 6.0}
	}
	return (hit*prices.hit + miss*prices.miss + output*prices.output) / 1_000_000
}

// mockRand 确定性伪随机，保证同一小时的值稳定（等价 Swift MockData.rand）。
func mockRand(time, salt int64, scale int) int {
	x := uint64(time) * 0x9E3779B97F4A7C15
	x += uint64(salt)
	x ^= x >> 30
	x *= 0xBF58476D1CE4E5B9
	x ^= x >> 27
	x *= 0x94D049BB133111EB
	x ^= x >> 31
	if scale <= 1 {
		return 0
	}
	return int(x % uint64(scale))
}

func fnv1a(s string) uint32 {
	var h uint32 = 2166136261
	for i := 0; i < len(s); i++ {
		h ^= uint32(s[i])
		h *= 16777619
	}
	return h
}

func intPtr(v int) *int       { return &v }
func strPtr(s string) *string { return &s }
func boolPtr(b bool) *bool    { return &b }

File: Sources/DeepSeekPanel/MockData.swift
import Foundation

/// 本地模拟数据：结构与真实接口一致，按小时生成、按天变化，
/// 用于离线测试，避免被平台 WAF 限流干扰。
enum MockData {
    static func fetch(
        window: StatsWindow,
        tz: Int
    ) -> (UserSummary, [APIKeyInfo], UsageAmountData, CostData) {
        let keys = [
            APIKeyInfo(
                createdAt: 1_700_000_000,
                lastUse: nil,
                trackingId: "mock-claude",
                sensitiveId: "sk-mock-1***",
                name: "claude"
            ),
            APIKeyInfo(
                createdAt: 1_700_000_000,
                lastUse: nil,
                trackingId: "mock-codex",
                sensitiveId: "sk-mock-2***",
                name: "codex"
            ),
            APIKeyInfo(
                createdAt: 1_700_000_000,
                lastUse: nil,
                trackingId: "mock-cursor",
                sensitiveId: "sk-mock-3***",
                name: "cursor"
            ),
            APIKeyInfo(
                createdAt: 1_700_000_000,
                lastUse: nil,
                trackingId: "mock-wechat",
                sensitiveId: "sk-mock-4***",
                name: "wechat filter"
            ),
        ]

        let models = [
            "deepseek-chat & deepseek-reasoner",
            "deepseek-v4-flash",
            "deepseek-v4-pro",
        ]

        var amountSeries: [AmountSeries] = []
        var cnySeries: [CostSeries] = []
        var usdSeries: [CostSeries] = []

        for key in keys {
            let keyModels: [String]
            switch key.name {
            case "claude":
                keyModels = ["deepseek-v4-flash", "deepseek-v4-pro"]
            case "codex":
                keyModels = ["deepseek-chat & deepseek-reasoner", "deepseek-v4-pro"]
            case "wechat filter":
                keyModels = ["deepseek-v4-flash"]
            default:
                continue // cursor 不活跃
            }

            for model in keyModels {
                var amountBuckets: [AmountBucket] = []
                var cnyBuckets: [CostBucket] = []
                var usdBuckets: [CostBucket] = []
                var time = window.requestStart
                while time < window.requestEnd {
                    let usage = usage(at: time, keyName: key.name, model: model)
                    amountBuckets.append(AmountBucket(time: time, usage: usage))
                    let usd = costUSD(usage, model: model)
                    cnyBuckets.append(
                        CostBucket(time: time, cost: String(format: "%.10f", usd * 7.2))
                    )
                    usdBuckets.append(
                        CostBucket(time: time, cost: String(format: "%.10f", usd))
                    )
                    time += 3600
                }

                let seriesKey = SeriesKey(
                    trackingId: key.trackingId,
                    name: key.name,
                    sensitiveId: key.sensitiveId,
                    valid: true
                )
                amountSeries.append(
                    AmountSeries(apiKey: seriesKey, model: model, buckets: amountBuckets)
                )
                cnySeries.append(CostSeries(apiKey: seriesKey, model: model, buckets: cnyBuckets))
                usdSeries.append(CostSeries(apiKey: seriesKey, model: model, buckets: usdBuckets))
            }
        }

        let amount = UsageAmountData(
            start: window.requestStart,
            end: window.requestEnd,
            bucket: 3600,
            models: models,
            series: amountSeries
        )
        let cost = CostData(
            start: window.requestStart,
            end: window.requestEnd,
            bucket: 3600,
            models: models,
            data: [
                CurrencyCostSeries(currency: "CNY", series: cnySeries),
                CurrencyCostSeries(currency: "USD", series: usdSeries),
            ]
        )
        let summary = UserSummary(
            normalWallets: [
                Wallet(currency: "USD", balance: "8.4200000000", tokenEstimation: "0"),
                Wallet(currency: "CNY", balance: "3.1400000000", tokenEstimation: "0"),
            ],
            bonusWallets: [
                Wallet(currency: "USD", balance: "0", tokenEstimation: "0"),
            ],
            totalCosts: [
                CostTotal(currency: "USD", amount: "0.8800000000"),
                CostTotal(currency: "CNY", amount: "1.2300000000"),
            ]
        )
        return (summary, keys, amount, cost)
    }

    private static func usage(at time: Int, keyName: String, model: String) -> TokenUsage {
        let hour = (time % 86400) / 3600
        let busy = (hour >= 9 && hour <= 22) ? 1.0 : 0.2
        let keyFactor = keyName == "wechat filter" ? 0.04 : 1.0
        let modelFactor = model == "deepseek-v4-pro" ? 2.0 : 1.0
        let base = Int(2_000_000 * busy * keyFactor * modelFactor)
        let salt = keyName.hashValue

        let hit = rand(time, salt: salt &+ 1, scale: base)
        let miss = rand(time, salt: salt &+ 2, scale: max(base / 8, 500))
        let response = rand(time, salt: salt &+ 3, scale: max(base / 20, 100))
        let requests = rand(time, salt: salt &+ 4, scale: max(base / 6000, 1))

        return TokenUsage(
            request: requests,
            responseToken: response,
            promptCacheHitToken: hit,
            promptCacheMissToken: miss
        )
    }

    private static func costUSD(_ usage: TokenUsage, model: String) -> Double {
        let hit = Double(usage.promptCacheHitToken ?? 0)
        let miss = Double(usage.promptCacheMissToken ?? 0)
        let output = Double(usage.responseToken ?? 0)
        let prices: (hit: Double, miss: Double, output: Double)
        if model == "deepseek-v4-pro" {
            prices = (0.025, 3.0, 6.0)
        } else {
            prices = (0.02, 1.0, 2.0)
        }
        return (hit * prices.hit + miss * prices.miss + output * prices.output) / 1_000_000
    }

    /// 确定性伪随机，保证同一小时的值稳定。
    private static func rand(_ time: Int, salt: Int, scale: Int) -> Int {
        var x = UInt64(truncatingIfNeeded: Int64(time)) &* 0x9E37_79B9_7F4A_7C15
        x = x &+ UInt64(truncatingIfNeeded: Int64(salt))
        x ^= x >> 30
        x = x &* 0xBF58_476D_1CE4_E5B9
        x ^= x >> 27
        x = x &* 0x94D0_49BB_1331_11EB
        x ^= x >> 31
        return Int(x % UInt64(max(scale, 1)))
    }
}
